### DISCUSSION TITLE
If I'm using monorepo with AdonisJS and don't want to publish my packages to npm, what's the best way to handle it?

### DISCUSSION THREAD

**Discussion by bosens-China** [adonisjs/core#4076] [Help]
I have several packages, but I don't want to publish them on npm. 
In Webpack, I can use "build" to combine files, but with AdonisJS, I need to use something like tsc to package the modules, which can result in the built resources still referencing the packages, such as "require('xxx/xxx')". 
Is there a good solution for my situation?

**Comment by RomainLanz** [comment]
Hey @bosens-China! 👋🏻 

I don't understand your issue.
Using a monorepo, you can install local packages easily without publishing them.

I recommend reading more about the monorepo tool you are using. You will find your answer in their documentation.

**Reply by bosens-China** [reply]
My meaning is that if I reference other local packages during the development phase, it's good to manage them through soft links in the development phase. However, when I try to build, there will be a problem because I don't want to publish other local packages to npm at all, which leads to the inability to find the relevant referenced package when running the files under the dist directory, such as the example below:
``` ts
import test from '@bo/test'
```
Compiled as:

``` js
const test  = require('@bo/test')
```

Note that at this point, I'm using Docker to copy the files under the dist directory and the package.json file, and then running npm install will fail.

**Comment by milesj** [comment]
Did you ever find a solution for this?

I'm trying to import TS files directly from a local package, but Adonis barfs on it.

**Reply by wswawan** [reply]
Try this the [npm link](https://docs.npmjs.com/cli/v10/commands/npm-link)

**Reply by milesj** [reply]
That's basically what workspaces are, but adonis tsc doesn't transform files in node modules.

**Comment by holoyan** [comment]
use [yalc](https://github.com/wclr/yalc)

or npm link as @wswawan mention

**Reply by milesj** [reply]
Ok but how does adonis know that these files need to be compiled? They are raw TS files.

**Reply by Julien-R44** [reply]
there's nothing to compile/transform dude, you just need to keep the local package in your node_modules once deployed in production

we usually use `pnpm deploy` for this https://pnpm.io/cli/deploy ( if you are using pnpm of course )

**Reply by milesj** [reply]
I don't think you're understanding the problem. This is development problem, not a deploy one. I want to import these packages WITHOUT a build step.

When I try and import the package (which are TS files) from within the Adonis app, the compilation literally fails with:

```
      "type": "SyntaxError",
      "message": "The requested module '@internal/validation/users' does not provide an export named 'loginSchema'",
      "stack":
          /api/app/procedures/auth.ts:1
          import { loginSchema } from '@internal/validation/users';
                   ^
```

Or:

```
      "type": "Error",
      "message": "Cannot find module '/api/node_modules/@internal/validation/users' imported from /api/app/procedures/auth.ts",
      "stack":
          Error: Cannot find module '/api/node_modules/@internal/validation/users' imported from /api/app/procedures/auth.ts
```

Where `@internal/validation/users'` maps to `/api/node_modules/@internal/validation/users.ts`.

I've tried changing a bunch of TS compiler options. Adding `package.json` imports/exports. Using project references. A bunch of stuff, nothing works correctly. These approaches all work fine in other apps, like Nuxt and Next, but not Adonis.

Adonis build process is weird. I'm puzzled why they don't just use a proper bundler? I was excited it had experimental vite support, but then was let down that it was only for frontend stuff.

**Reply by Julien-R44** [reply]
There's nothing "weird" about the Adonis build process. in fact, it's totally standard, without anything fancy: simply Node with ts-node as a loader for transpiling, nothing more. 
Why should we use Vite or a bundler for backend development? It adds a layer of complexity that is often unnecessary and can cause problems.

I think it would be great if you took some time to think about why things were done the way they were before criticizing freely without any basis.

--- 

That said. The problem you're having is probably related to the ESM/CJS interop. Add a "type": "module" to your @internal/validation package and it should be good.

In other words: the problem you're experiencing has nothing to do with Adonis.

If you're having trouble, please make a repro and I'll take a look.

**Reply by holoyan** [reply]
>Ok but how does adonis know that these files need to be compiled? They are raw TS files.

@milesj if I understood you correct then I had similar question while developing [this package](https://github.com/holoyan/adonisjs-permissions/tree/dev) (it's still under development) 

let me share my experince, this maybe will help you. So I started using [starter kit](https://github.com/adonisjs/pkg-starter-kit) skeleton
and after some development I wanted to install it in my adonis project (local), to do that I used yalc, but for some reason I faced an [issue](https://github.com/adonisjs/core/discussions/4432) then I switched to `npm link` and now it works good, I just build and that's it


To answer issue about exports

>I've tried changing a bunch of TS compiler options. Adding package.json imports/exports. Using project references. A bunch of stuff, nothing works correctly. These approaches all work fine in other apps, like Nuxt and Next, but not Adonis.

You should write all your `exportable` classes,functions inside [index.ts](https://github.com/holoyan/adonisjs-permissions/blob/dev/index.ts) file, if you look into package.json [export](https://github.com/holoyan/adonisjs-permissions/blob/dev/package.json#L19) property, there is defined all exportable files,
and you should specify list of [files](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#files) that you want to *publish* during install, this is in [my case](https://github.com/adonisjs/pkg-starter-kit/blob/main/package.json#L9) 
And of course don't forget to update [configute.ts file](https://github.com/holoyan/adonisjs-permissions/blob/dev/configure.ts), 
From the docs
>The configure hook is called when someone runs "node ace configure <package>"
| command. You are free to perform any operations inside this function to
| configure the package.


Hope this was helpfull
